### PR TITLE
Gaurav's Attandence feature

### PR DESCRIPTION
**Pull request by gauravThakur-sgl**
# Featured Added
### Punch In Functionality
    Allows users to mark their punch-in time
    Disables the "Punch In" button after successful punch-in.
### Punch Out Functionality

    Allows users to mark their punch-out time
    Calculates the total time between punch-in and punch-out.
    Validates whether the user has completed 9 hours:
    Displays an error message if 9 hours are not completed.
    Marks attendance as completed upon successful validation.
    Enables the "Punch In" button for the next session after successful punch-out.

### Error Handling 
    Catches and logs errors during punch-in or punch-out operations to ensure robustness.

### UI Enhancements
    Displays punch-in time or success message in a styled card.
    Provides clear visual feedback for disabled/enabled buttons based on the punch-in/punch-out state.

### CODE REVIEW

**Comment by gauravThakur-sgl** [comment]
A feature for viewing Punching History has been introduced. Additionally, redux-persist has been implemented to ensure state persistence.

File: gaurav/src/redux/reducers.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      return {
+        ...state,
+        punchInDate: action.payload,
+        punchData: [...state.punchData, { time: action.payload }]
+      };
+    case "PUNCH_OUT":
+      return {
+        ...state,
+        punchOutDate: action.payload,
+        punchData: [...state.punchData, { time: action.payload }]
```
Comment (shubhamships) [inline gaurav/src/redux/reducers.ts:26]: Can you add an example data for the contents of `punchData`

Comment (gauravThakur-sgl) [reply gaurav/src/redux/reducers.ts:26]: when user `punchIn`  or `punchOut` then that time will be added to the `punchData []` array.
and then we `map` the `punchData[]` to show the `punchIn` and `punchOut` timew based on the `index.`

Screenshot from 2024-11-26 23-00-12

Comment (shubhamships) [reply gaurav/src/redux/reducers.ts:26]: I meant the data in JSON format @gauravThakur-sgl

Comment (DJ-LIFE) [reply gaurav/src/redux/reducers.ts:26]: punchData in JSON format : 

Screenshot from 2024-11-27 16-07-20

**Review by shubhamships — APPROVED** [review]
Good to merge

**Pull request merged**